oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/NimbusJwtClientAuthenticationParametersConverter.java

```diff
@@ -0,0 +1,217 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.client.endpoint;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.JWKSet;
+import com.nimbusds.jose.jwk.KeyType;
+import com.nimbusds.jose.jwk.source.ImmutableJWKSet;
+import com.nimbusds.jose.jwk.source.JWKSource;
+import com.nimbusds.jose.proc.SecurityContext;
+
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.core.ClientAuthenticationMethod;
+import org.springframework.security.oauth2.core.OAuth2AuthorizationException;
+import org.springframework.security.oauth2.core.OAuth2Error;
+import org.springframework.security.oauth2.core.endpoint.OAuth2ParameterNames;
+import org.springframework.security.oauth2.jose.jws.JwsAlgorithm;
+import org.springframework.security.oauth2.jose.jws.MacAlgorithm;
+import org.springframework.security.oauth2.jose.jws.SignatureAlgorithm;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.util.Assert;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+
+/**
+ * A {@link Converter} that customizes the OAuth 2.0 Access Token Request parameters by
+ * adding a signed JSON Web Token (JWS) to be used for client authentication at the
+ * Authorization Server's Token Endpoint. The private/secret key used for signing the JWS
+ * is supplied by the {@code com.nimbusds.jose.jwk.JWK} resolver provided via the
+ * constructor.
+ *
+ * <p>
+ * <b>NOTE:</b> This implementation uses the Nimbus JOSE + JWT SDK.
+ *
+ * @param <T> the type of {@link AbstractOAuth2AuthorizationGrantRequest}
+ * @author Joe Grandja
+ * @since 5.5
+ * @see Converter
+ * @see com.nimbusds.jose.jwk.JWK
+ * @see JwtCustomizer
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7523#section-2.2">2.2
+ * Using JWTs for Client Authentication</a>
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7521#section-4.2">4.2
+ * Using Assertions for Client Authentication</a>
+ * @see <a target="_blank" href="https://connect2id.com/products/nimbus-jose-jwt">Nimbus
+ * JOSE + JWT SDK</a>
+ */
```

Comment:
Can you add documentation what this is expected to be injected into?

Reply:
I plan on adding documentation and sample code in the reference after this PR is accepted and before I merge.

Reply:
I think something on the Javadoc is worthwhile here. At minimum an `@see` to the object that consumes this.

Reply:
I added `@see` for `OAuth2AuthorizationCodeGrantRequestEntityConverter` and `OAuth2ClientCredentialsGrantRequestEntityConverter` - likely the ones that will be used the most